### PR TITLE
Feat: Share button in zone header

### PR DESCRIPTION
**Pull request by cadeban**
## Description

[AVO-480](https://linear.app/electricitymaps/issue/AVO-480/add-zonecountrymap-share-button)
tbd

### Preview

tbd

### Double check

- [ ] I have tested my parser changes locally with `poetry run test_parser "zone_key"`
- [ ] I have run `pnpx prettier@2 --write .` and `poetry run format` in the top level directory to format my changes.

### CODE REVIEW

**Comment by cadeban** [comment]
@tonypls I'm still polishing it up & wrapping up testing, but would you mind skimming this when you have a chance? Check that it fits with the historical linking work?

**Review by tonypls — COMMENTED** [review]
Visual suggestions:
It's super wide on desktop and the text could be more meaningful:
image
Perhaps "URL copied to clipboard!"

I think the right align of the icon should line up with the vertical line in this image and the icon should be center aligned with the HeaderTitle component, there's also some extra spacing between the title and date too.
image

I also think it would be super nice to use a cypress component test here, the unit test is nice but for example if I make it so the toast no longer closes the tests will still pass, a cypress component test will enable us to test these interactions and visually debug. 

Code looks nice! Will dig a bit deeper after the alignment changes.

File: web/src/features/map-controls/MapControls.tsx [hunk 1]
```diff
@@ -109,7 +109,7 @@ function DesktopMapControls() {
   };
 
   return (
-    <div className="pointer-events-none absolute right-3 top-2 z-30 hidden flex-col items-end md:flex">
```
Comment (tonypls) [inline web/src/features/map-controls/MapControls.tsx:112]: This change will mean the buttons are hidden behind the legend if you have the screen in a ridiculous size, unlikely case but think it's good to leave it at 30

File: web/src/features/panels/zone/ShareButton.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
+import { Button, ButtonProps } from 'components/Button';
+import { Toast, useToastReference } from 'components/Toast';
+import { isiOS, isMobile } from 'features/weather-layers/wind-layer/util';
```
Comment (tonypls) [inline web/src/features/panels/zone/ShareButton.tsx:3]: Would be nice to try not use the windlayer stuff here if possible. 

For mobile I'd love to unify all the checks but now isn't the time so I think this is the most up to date one `useIsMobile`.

For iOS check we kinda have one in here isOldIOSVersion that we could extract into a util that isn't in the wind layer.

**Comment by VIKTORVAV99** [comment]
> Visual suggestions:
> 
> It's super wide on desktop and the text could be more meaningful:
> 
> image
> 
> Perhaps "URL copied to clipboard!"
> 
> 
> 
> I think the right align of the icon should line up with the vertical line in this image and the icon should be center aligned with the HeaderTitle component, there's also some extra spacing between the title and date too.
> 
> image
> 
> 
> 
> I also think it would be super nice to use a cypress component test here, the unit test is nice but for example if I make it so the toast no longer closes the tests will still pass, a cypress component test will enable us to test these interactions and visually debug. 
> 
> 
> 
> Code looks nice! Will dig a bit deeper after the alignment changes.
> 
> 
> 
> 

Didn't we change the Figma so the share button are center aligned with the back arrow?

**Comment by tonypls** [comment]
> Visual suggestions: It's super wide on desktop and the text could be more meaningful: image Perhaps "URL copied to clipboard!"
> 
> I think the right align of the icon should line up with the vertical line in this image and the icon should be center aligned with the HeaderTitle component, there's also some extra spacing between the title and date too. image
> 
> I also think it would be super nice to use a cypress component test here, the unit test is nice but for example if I make it so the toast no longer closes the tests will still pass, a cypress component test will enable us to test these interactions and visually debug.
> 
> Code looks nice! Will dig a bit deeper after the alignment changes.

I was on an older preview 🤦 disregard the layout suggestions!

File: web/src/features/panels/zone/ShareButton.tsx (hunk 2 above)
Comment (cadeban) [reply web/src/features/panels/zone/ShareButton.tsx:3]: Agreed, it's definitely time to extract this

**Comment by cadeban** [comment]
To summarize, the button layout is ok in the most recent branch deployment, but these are items to add/fix:
    - [ ] Cypress Toast test
    - [ ] Fix banner length on desktop
    - [ ] Update banner text
    - [ ] Extract mobile logic
If it's alright with y'all, I may make a follow up PR with the mobile logic, since it touches other files. [Linear issue](https://linear.app/electricitymaps/issue/AVO-552/extract-ismobileisiphoneisandroid-etc-logic-into-own-utilhook)

**Comment by VIKTORVAV99** [comment]
> To summarize, the button layout is ok in the most recent branch deployment, but these are items to add/fix:
> 
> * [ ]  Cypress Toast test
> * [x]  Fix banner length on desktop
> * [x]  Update Toast text - to revisit how to handle
> * [x]  Extract isMobile etc. logic
> 
> If it's alright with y'all, I may make a follow up PR with the isMobile etc. logic, since it touches other files/I would like to add tests. [Linear issue](https://linear.app/electricitymaps/issue/AVO-552/extract-ismobileisiphoneisandroid-etc-logic-into-own-utilhook)
> 
> TODO:
> 
> * [ ]  toast translations

For me it would be fine with a follow up PR that revises the internal logic.
I don't know what you had planned but I'd look into making these either hooks or even atoms and extract the logic from react completely.

**Comment by VIKTORVAV99** [comment]
> To summarize, the button layout is ok in the most recent branch deployment, but these are items to add/fix:
> 
> * [ ]  Cypress Toast test
> * [x]  Fix banner length on desktop
> * [x]  Update Toast text - to revisit how to handle
> * [x]  Extract isMobile etc. logic
> 
> If it's alright with y'all, I may make a follow up PR with the isMobile etc. logic, since it touches other files/I would like to add tests. [Linear issue](https://linear.app/electricitymaps/issue/AVO-552/extract-ismobileisiphoneisandroid-etc-logic-into-own-utilhook)
> 
> TODO:
> 
> * [ ]  toast translations

BTW if you add these todos to the PR description instead github will keep track of the progress in overviews, a nifty little feature :)

File: web/src/features/panels/zone/ZoneHeaderTitle.tsx [hunk 3]
```diff
@@ -71,12 +70,15 @@ export default function ZoneHeaderTitle({ zoneId }: ZoneHeaderTitleProps) {
...
             <TooltipWrapper side="bottom" tooltipContent={disclaimer}>
-              <Info className="ml-auto shrink-0 text-gray-500" />
+              <Info className="text-gray-500" />
             </TooltipWrapper>
           )}
         </div>
         <TimeDisplay className="whitespace-nowrap text-sm" />
       </div>
+      <div className="ml-auto self-center px-3">
+        <ShareButton />
```
Comment (cadeban) [inline web/src/features/panels/zone/ZoneHeaderTitle.tsx:80]: > Is this div really needed?
> 
> I'd think you can add the classes directly to the button?
> Or even extract the alignment to the outer flexbox.

copied from https://github.com/electricitymaps/electricitymaps-contrib/pull/7248/files#r1781130997

**Comment by cadeban** [comment]
> I don't know what you had planned but I'd look into making these either hooks or even atoms and extract the logic from react completely.

So I've been thinking a little bit about the isIphone/isAndroid/etc functions. Could you share a little bit more about the benefit of making them atoms or hooks? My thinking is that utility functions can be used in more contexts than atoms/hooks. Is it about the cost of repeatedly calculating them?

I do think we can clean up the logic potentially, though.

**Comment by VIKTORVAV99** [comment]
> > I don't know what you had planned but I'd look into making these either hooks or even atoms and extract the logic from react completely.
> 
> 
> 
> So I've been thinking a little bit about the isIphone/isAndroid/etc functions. Could you share a little bit more about the benefit of making them atoms or hooks? My thinking is that utility functions can be used in more contexts than atoms/hooks. Is it about the cost of repeatedly calculating them?
> 
> 
> 
> I do think we can clean up the logic potentially, though.

Yeah it's mostly about reducing the calculations. I'm not sure they are very costly though.

File: web/src/features/panels/zone/ShareButton.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+  };
+
+  const copyToClipboard = async () => {
+    try {
+      await navigator.clipboard.writeText(url);
+      setToastMessage('URL copied to clipboard!');
+    } catch (error) {
+      console.error(error);
+      setToastMessage('Error copying URL to clipboard');
```
Comment (VIKTORVAV99) [inline web/src/features/panels/zone/ShareButton.tsx:56]: Just don't forget to add translations for these.

Don't repeat my mistake... 😅

File: web/src/components/WhatsNewTooltip.tsx [hunk 5]
```diff
@@ -0,0 +1,5 @@
+// interface WhatsNewTooltipProps {}
```
Comment (tonypls) [inline web/src/components/WhatsNewTooltip.tsx:1]: Is this used?

File: web/src/features/panels/zone/ShareButton.tsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  shareUrl?: string;
+  showiOSIcon?: boolean;
+}
+const trackShareClick = trackShare(ShareType.SHARE);
+const DURATION = 3 * 1000;
+
+export function ShareButton({
+  iconSize = DEFAULT_ICON_SIZE,
+  shareUrl,
+  showiOSIcon = isiOS(),
```
Comment (tonypls) [inline web/src/features/panels/zone/ShareButton.tsx:26]: Nitpick but I think consistency is worth the ugliness in this case.
```suggestion
  showIosIcon = isIos(),
```

**Review by tonypls — APPROVED** [review]
LGTM!! Super nice with good test coverage. 🎉 

Couple of small comments but happy if you want to merge them in and I'll change them when I'm adding the historic link

File: web/src/components/WhatsNewTooltip.tsx (hunk 5 above)
Comment (cadeban) [reply web/src/components/WhatsNewTooltip.tsx:1]: 🤦

File: web/src/features/panels/zone/ShareButton.tsx (hunk 6 above)
Comment (tonypls) [reply web/src/features/panels/zone/ShareButton.tsx:26]: I'm very pro releasing this on turning the feature flag on today 🚢

File: web/src/components/WhatsNewTooltip.tsx (hunk 5 above)
Comment (cadeban) [reply web/src/components/WhatsNewTooltip.tsx]: Removed, must've forgotten to stash it

File: web/src/features/panels/zone/ShareButton.tsx (hunk 6 above)
Comment (cadeban) [reply web/src/features/panels/zone/ShareButton.tsx]: Fixed - will merge it in when everything passes. Thanks for the review!

**Pull request merged**